### ISSUE TITLE
[Feature Request] Automation - allow switch of algorithm oref0/oref1

### ISSUE THREAD

**Issue by dansuk** [MilosKozak/AndroidAPS#2012]
Incorporate the ability to switch between oref0/AMA, and oref1/SMB in the automation options to allow a highly flexible system.

For most of the time I find that OpenAPS AMA works best for me as the initial period of zero temp with oref1 causes highs (this seems to be experienced by the wider community too from forum posts)  but OpenAPS SMB clearly is more useful on occassions with the UAM and ability to bolus to recover from high levels quicker.

A system which allows the flexibility to automatically choose the algorithm depending on the cicrcumstances would benefit everyone.

Common situations could be:
If blood glucose, raises above >12 activate oref1/SMB (allow SMB bolusing).
If location is xxx (where I know that i will struggle with carb counting/buffet food) then active oref1/SMB
If COB=0 then activate oref0/AMA
If blood glucose<8 then activate oref0/AMA

**Comment by thomasmenges** [comment]
I can only agree with you and find this extension very helpful. I my opinion that would be a  great point for all lopper who are not using SMB.

For me the automation plugin also lacks the possibility to trigger a bolus (SMB 0,1) if certain events occur.

**Comment by winni67** [comment]
Same to me. I use OpenAMA, because of the zero temp highs and the slow increase of insulin for high carb meals. Setting eating soon temp targets or a smal meal bolus 1h before meal are to risky for me. But perhaps in SMB, it should be adjustable, what happens the two hours after meal bolus. Working like AMA/ref0 at this part (without the 2h zero temp) would help me. The rest of the day, SMB/ref1 would be preferred.

**Comment by PoweRGbg** [comment]
@dansuk 
First you turn off SMB Always, SMB After carbs and and SMB with high TT
If blood glucose, raises above >12 activate oref1/SMB (allow SMB bolusing). -> Automation trigger should be "Glucose greater than 12 "-> Action should be "Start temp target of 4.5 for 30 minutes". This should activate SMB bolusing. 
If location is xxx (where I know that i will struggle with carb counting/buffet food) then active oref1/SMB -> Same as above but trigger should be the location.
If COB=0 then activate oref0/AMA -> If COB=0 and the SMB options turned off  then SMB acts as AMA. Isn't it?
If blood glucose<8 then activate oref0/AMA -> Same as above.

So I think that the enhancement you want is already achievable with the current version.

**Comment by thomasmenges** [comment]
@PowerRGbg
I think that the enhancement isn't available in the current version.
You have to decide manually if you're using AMA or SMB.
The automation plugin isn't able to change from AMA to SMB or in the other direction automatically.
That is the target of this issue / enhancement.

**Comment by PoweRGbg** [comment]
@thomasmenges 
Please be specific which of the recipes I listed above is not working or what you don't like about it ? If this issue is about "I want AMA, not SMB!" , then there is a switch in the Config builder. Otherwise I believe you could force SMB to act as AMA via setting and cancelling temp targets.

**Comment by dansuk** [comment]
@PoweRGbg thanks for the idea. I am testing this now however oref1/SMB in this set up appears to still do the calculation differently - see my current situation:

Right now I have just eaten, I have a blood glucose of 7,6, 30 COB (no temp-target set). The oref1/SMB algorithm is enforcing a zero temp basal. If I cancel this manually, it again implements a zero temp basal on the next itteration. The minPregBG is shown as 3.2, minGuardBG 2.3, IOBpredBG, 2.2 COB, PredBG10.8, UA pred BG 2.2, minGuardBG 2.3<3.9 .

In the same loop itteration, when I switch to oref0/AMA there is no zero temp basal as eventual BG 8.8 >=5,5 resulting in a 300% basal temp - quite a difference.

**Comment by PoweRGbg** [comment]
@dansuk  I fully understand because I had similar issues with SMB. In your case the only predicted high is based on COB and I suppose your BG of 7.6 is lower or equal to the one before it. That is the reason to have all other predicted values to be low and to have a zero temp set.
Many other users are dealing huge spikes with SMB also and they prefer to  suspend loop during mealtime. Setting a low TT (4.5 mmol) 30-60 min before meal helps a lot. It is like prebolusing. It prevents or lowers the posprandial spike. Another way to deal with such issues is to lower the carb ratio. Setting TT or switching profile percentage can also be done via Automation. Good luck!

**Comment by dansuk** [comment]
@PoweRGbg yes, I am daily trying to fine-tune things with SMB and follow all the suggestions on the forums and hoping sometime I will be able to nail the setting. I will keep working on it and would rather not have to suspend anything. I still find both implementations of oref very useful in different situations and I was hoping providing the option in automation would allow the best of both without having to do so much manually, empower the user a bit more with the flavour depending on the circumstances. Hoping it would not be too difficult to add an option (even looked at the code myself in desperation lol but its beyond me!)...

**Comment by tynbendad** [comment]
i followed PowerRGbg's instructions above (https://github.com/MilosKozak/AndroidAPS/issues/2012?fbclid=IwAR3cKg2H7v9qE7Upt4aIx3WRR1bfzoPUrYx9HL_vKI0rOl3XyZFNBIgSl6U#issuecomment-537829577) and it sort of works. there are two problems with it though:
1) it uses the temp target, so you can no longer use it for manual temp targets.  this is sort of ok, except you have to learn/teach a new thing to replace the old, and the replacement (profile switching) doesn't fully work thru nightscout yet since you can't give it a percentage thru the dialog box yet (a separate issue for nightscout).
2) like the original suggestion in this issue, it allows you to turn smb on and off, but doesn't allow you to adjust smb minutes from the automation rules.

fwiw, if i could do it (i can't), instead of the original request or PowerRGbg's workaround, i would make aaps changes to:
A) directly set smb minutes from the automation rules
B) fix/add SMB UAM minutes to match at least openaps 0.6.2 (currently aaps UAM is using a fixed 30 minute SMB instead of SMB minutes setting).

**Comment by tynbendad** [comment]
fwiw these screenshots show how i setup @PoweRGbg 's workaround. (i'm no longer using this as i've hacked aaps locally [which i believe is generally discouraged.])
![Screenshot_2019-11-29-14-35-54-780_info nightscout androidaps](https://user-images.githubusercontent.com/20666529/69891254-5fdfb400-12b8-11ea-88ee-d82a2ed528e1.jpg)
![IMG_20191129_145038](https://user-images.githubusercontent.com/20666529/69891259-65d59500-12b8-11ea-923f-a370bb72acae.jpg)
![IMG_20191129_144207](https://user-images.githubusercontent.com/20666529/69891261-69691c00-12b8-11ea-8a1b-3c9251ee852f.jpg)
